app/lib/features/pins/widgets/pin_item.dart

```diff
@@ -107,90 +122,268 @@ class _PinItemState extends ConsumerState<PinItem> {
     );
   }
 
-  // pin content widget
-  Widget _buildPinDescription() {
+// attachment list UI
+  Widget _buildAttachmentList() {
+    final noAttachmentTextStyle = Theme.of(context)
+        .textTheme
+        .labelMedium!
+        .copyWith(color: Theme.of(context).colorScheme.neutral5);
+    final attachmentTitleTextStyle = Theme.of(context).textTheme.labelLarge;
     final pinEdit = ref.watch(pinEditProvider(widget.pin));
-    final pinEditNotifier = ref.watch(pinEditProvider(widget.pin).notifier);
+    final attachments = ref.watch(pinAttachmentsProvider(widget.pin));
+    final selectedAttachments = ref.watch(selectedPinAttachmentsProvider);
+    final attachmentNotifier =
+        ref.watch(selectedPinAttachmentsProvider.notifier);
+    return attachments.when(
+      data: (list) {
+        return Column(
+          crossAxisAlignment: CrossAxisAlignment.start,
+          children: <Widget>[
+            Padding(
+              padding: const EdgeInsets.all(8),
+              child: Row(
+                children: [
+                  Text('Attachments', style: attachmentTitleTextStyle),
+                  const SizedBox(width: 5),
+                  const Icon(Atlas.paperclip_attachment_thin, size: 14),
+                  const SizedBox(width: 5),
+                  Text('${list.length}'),
+                ],
+              ),
+            ),
+            Wrap(
+              spacing: 5.0,
+              runSpacing: 10.0,
+              children: <Widget>[
+                if (list.isNotEmpty)
+                  for (var item in list)
+                    AttachmentTypeHandler(
+                      attachment: item,
+                      pin: widget.pin,
+                    ),
+                for (var item in selectedAttachments)
+                  Stack(
+                    children: <Widget>[
+                      AttachmentContainer(
+                        pin: widget.pin,
+                        filename: item.file.path.split('/').last,
+                        child: attachmentIconHandler(item.file, 20),
+                      ),
+                      Visibility(
+                        visible: pinEdit.editMode,
+                        child: Positioned(
+                          top: -15,
```

Comment:
I feel this is nesting too deep and should probably be outsourced into a local sub-build-widget-rendering-function...